examples/shader/array_texture.rs

```diff
@@ -0,0 +1,170 @@
+use bevy::{
+    ecs::system::{lifetimeless::SRes, SystemParamItem},
+    pbr::MaterialPipeline,
+    prelude::*,
+    reflect::TypeUuid,
+    render::{
+        render_asset::{PrepareAssetError, RenderAsset, RenderAssets},
+        render_resource::{
+            AddressMode, BindGroup, BindGroupDescriptor, BindGroupEntry, BindGroupLayout,
+            BindGroupLayoutDescriptor, BindGroupLayoutEntry, BindingResource, BindingType,
+            SamplerBindingType, ShaderStages, TextureSampleType, TextureViewDimension,
+        },
+        renderer::RenderDevice,
+    },
+};
+
+/// This example illustrates how to create a texture for use with a texture2DArray shader uniform
+/// variable.
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .add_plugin(MaterialPlugin::<ArrayTextureMaterial>::default())
+        .add_startup_system(setup)
+        .add_system(create_array_texture)
+        .run();
+}
+
+fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
+    // Start loading the texture.
+    commands.insert_resource(LoadingTexture(Some(
+        asset_server.load("textures/array_texture.png"),
+    )));
+
+    // camera
+    commands.spawn_bundle(PerspectiveCameraBundle {
+        transform: Transform::from_xyz(2.0, 2.0, 2.0).looking_at(Vec3::ZERO, Vec3::Y),
+        ..Default::default()
+    });
+}
+
+struct LoadingTexture(Option<Handle<Image>>);
+
+fn create_array_texture(
+    mut commands: Commands,
+    mut loading_texture: ResMut<LoadingTexture>,
+    mut textures: ResMut<Assets<Image>>,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<ArrayTextureMaterial>>,
+) {
+    let (handle, texture) = match loading_texture.0.as_ref() {
+        Some(handle) => {
+            if let Some(texture) = textures.get_mut(handle) {
+                (loading_texture.0.take().unwrap(), texture)
+            } else {
+                return;
+            }
+        }
+        None => return,
+    };
+
+    // Create a new array texture asset from the loaded texture.
+    let array_layers = 4;
+    texture.reinterpret_stacked_2d_as_array(array_layers);
+    texture.sampler_descriptor.address_mode_u = AddressMode::Repeat;
+    texture.sampler_descriptor.address_mode_v = AddressMode::Repeat;
+
+    // Spawn a cube that's shaded using the array texture.
+    commands.spawn().insert_bundle(MaterialMeshBundle {
+        mesh: meshes.add(Mesh::from(shape::Cube { size: 1.0 })),
+        material: materials.add(ArrayTextureMaterial {
+            array_texture: handle,
+        }),
+        ..Default::default()
+    });
+}
```

Comment:
I don't quite get this code. I assume the logic is supposed to be run only once, but the way it is right now wouldn't it be executed every frame once the texture is loaded?

Should there be a `*loading_texture.0 = None;` at the bottom?

Reply:
It looks like there was a `.take()` called on the `Option<Handle<Image>>`. But I reworked it anyway.